File: 04/haskell/main.hs
import Data.List
import Data.List.Split (chunksOf, splitOn)
import Debug.Trace
import System.IO

type BingoBoard = [[Int]]

isWinner :: BingoBoard -> Bool
isWinner board = any (all (-1 ==)) board || any (all (-1 ==)) (transpose board)

markBoard :: Int -> BingoBoard -> BingoBoard
markBoard val board = [[z | x <- row, let z = if x /= val then x else -1] | row <- board]

scoreBoard :: BingoBoard -> Int -> Int
scoreBoard board val = val * sum (filter (-1 /=) $ concat board)

runSolution :: [BingoBoard] -> [Int] -> Int
runSolution boards (x : xs) = case winner of
  Nothing -> runSolution newBoards xs
  Just board -> scoreBoard board x
  where
    newBoards = map (markBoard x) boards
    winner = find isWinner newBoards
runSolution _ _ = undefined -- Should not happen

runSolution2 :: [BingoBoard] -> [Int] -> Int
runSolution2 boards (x : xs) = case winner of
  Nothing -> runSolution2 filteredBoards xs
  Just board -> if null filteredBoards then scoreBoard board x else runSolution2 filteredBoards xs
  where
    newBoards = map (markBoard x) boards
    winner = find isWinner newBoards
    filteredBoards = filter (not . isWinner) newBoards
runSolution2 _ _ = undefined -- Should not happen

solution :: [String] -> Int
solution input = runSolution2 boards nums
  where
    nums = map read $ splitOn "," $ head input
    boards = map (drop 1 . map (map read . words)) (chunksOf 6 $ drop 1 input)

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print (solution $ lines contents)
  hClose handle

File: 04/python/main.py
def is_winner(board):
    for row in board:
        if row == [-1] * 5:
            return True
    for i in range(0, 5):
        vals = [row[i] for row in board]
        if vals == [-1] * 5:
            return True
    # if board[0][0] == board[1][1] == board[2][2] == board[3][3] == board[4][4]:
    #     return True
    # if board[4][0] == board[3][1] == board[2][2] == board[1][3] == board[0][4]:
    #     return True
    return False
def solution(lines):
    nums = [int(x) for x in lines[0].split(',') if x]

    boards = []
    for i in range(2, len(lines), 6):
        board = [[int(z) for z in x.strip().split(' ') if z] for x in lines[i:i+5]]
        boards.append(board)

    count_won = 0

    for num in nums:
        for k in range(0, len(boards)):
            board = boards[k]
            if (board is not None):
                
                for i in range(0, len(board)):
                    for j in range(0, len(board[i])):
                        if board[i][j] == num:
                            board[i][j] = -1
                if (is_winner(board)):
                    print("----(winner)")
                    for row in board:
                        print(row)
        
                    boards[k] = None
                    count_won += 1

                    if (count_won == len(boards)):
                        s = 0
                        for row in board:
                            for val in row:
                                if val != -1:
                                    s += val
                        print(num * s)
                        return
                else:
                    print("----")
                    for row in board:
                        print(row)
        

    
    
    # print (lines)


with open('input.txt') as f:
    lines = f.readlines()
    print(solution(lines))